### PR TITLE
Enhance caliptra-api  to make it a suitable Rust replacement of  libcaliptra. 

### PR DESCRIPTION
**Pull request by rusty1968**
-  Introduce the SocManager super-trait by extracting a subset of re-usable methods from the HwModel trait.
 -  Moved the SocManager super-trait to the caliptra-api crate.
 -  Refactored SocManager super-trait to eliminate its dependency on the std crate, making it compatible with no_std 
     environments.
-  Modified tests to accomodate API differences resulting from the refactor.
-  Added a prelude to the caliptra-api.

### CODE REVIEW

**Comment by zhalvorsen** [comment]
Would you be willing to split the `no_std` refactor into its own commit? I think it would make it easier to review that part of the code. It can be hard to follow the changes right now because they're spread out across different files.

File: hw-model/src/lib.rs [hunk 1]
```diff
@@ -871,162 +706,92 @@ pub trait HwModel {
...
     /// the uC responded with data, `Ok(None)` if the uC indicated success
     /// without data, Err(ModelError::MailboxCmdFailed) if the microcontroller
     /// responded with an error, or other model errors if there was a problem
     /// communicating with the mailbox.
-    fn mailbox_execute(
+    fn mailbox_execute<'r>(
+        &mut self,
+        cmd: u32,
+        buf: &[u8],
+        resp_data: &'r mut MboxBuffer,
```
Comment (korran) [inline hw-model/src/lib.rs:726]: Should probably just prefer `&mut [u8]` here instead of a fixed-size type in the API.

Comment (zhalvorsen) [reply hw-model/src/lib.rs:739]: +1

Comment (rusty1968) [reply hw-model/src/lib.rs]: Done.

File: hw-model/src/lib.rs [hunk 2]
```diff
@@ -300,6 +292,48 @@ pub enum ModelError {
     },
     MailboxRespInvalidFipsStatus(u32),
     MailboxTimeout,
+    NoRequestsAvail,
+}
+
+impl From<CaliptraApiError> for ModelError {
```
[the remark this answers is not in the document]
Comment (rusty1968) [reply hw-model/src/lib.rs:313]: One of them needs to be able to implement std::Display so this was the least intrusive way to make this change.

[the remark this answers is not in the document]
Comment (zhalvorsen) [reply hw-model/src/lib.rs:313]: You should be able to implement `core::fmt::Display` in the `no_std` version

File: api/src/lib.rs [hunk 3]
```diff
@@ -1,11 +1,457 @@
...
+    /// If the cptra_fuse_wr_done has already been written, or the
+    /// hardware prevents cptra_fuse_wr_done from being set.
+    fn init_fuses(&mut self, fuses: &Fuses) {
+        if !self.soc_ifc().cptra_reset_reason().read().warm_reset() {
+            assert!(
+                !self.soc_ifc().cptra_fuse_wr_done().read().done(),
+                "Fuses are already locked in place (according to cptra_fuse_wr_done)"
+            );
+        }
+        //        println!("Initializing fuses: {:#x?}", fuses);
```
[the remark this answers is not in the document]
Comment (rusty1968) [reply api/src/lib.rs]: Done

File: hw-model/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (rusty1968) [reply hw-model/src/lib.rs:313]: The reason to implement Display is for Models to write text messages with the error diagnostics to the console. We won;t be doing that in an embedded environment.
impl Display for ModelError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            ModelError::MailboxCmdFailed(err) => write!(f, "Mailbox command failed. fw_err={err}"),
            ModelError::UnableToLockMailbox => write!(f, "Unable to lock mailbox"),
            ModelError::BufferTooLargeForMailbox => write!(f, "Buffer too large for mailbox"),
            ModelError::UploadFirmwareUnexpectedResponse => {
                write!(f, "Received unexpected response after uploading firmware")
            }

File: api/types/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+    }
+}
+
+#[derive(Copy, Clone, Debug, Default, PartialEq, Eq)]
+pub enum ErrorInjectionMode {
+    #[default]
+    None,
+    IccmDoubleBitEcc,
+    DccmDoubleBitEcc,
+}
```
[the remark this answers is not in the document]
Comment (rusty1968) [reply api/types/src/lib.rs]: Done

File: runtime/tests/runtime_integration_tests/main.rs [hunk 5]
```diff
@@ -5,7 +5,7 @@ mod test_authorize_and_stash;
 mod test_boot;
 mod test_certify_key_extended;
 mod test_certs;
-mod test_disable;
+//mod test_disable;
```
[the remark this answers is not in the document]
Comment (rusty1968) [reply runtime/tests/runtime_integration_tests/main.rs]: That was unintended . Done,. I reverted all changes to tests and then added only the  "use SocManager"  to minimize changes.

File: test/tests/fips_test_suite/security_parameters.rs [hunk 6]
```diff
@@ -154,6 +154,8 @@ pub fn attempt_ssp_access_rom() {
 #[test]
 #[cfg(not(feature = "test_env_immutable_rom"))]
 pub fn attempt_ssp_access_fw_load() {
+    //use dpe::response;
+
```
[the remark this answers is not in the document]
Comment (rusty1968) [reply test/tests/fips_test_suite/security_parameters.rs]: Done

File: test/tests/caliptra_integration_tests/main.rs [hunk 7]
```diff
@@ -1,6 +1,6 @@
 // Licensed under the Apache-2.0 license
 
-mod fake_collateral_boot_test;
+//mod fake_collateral_boot_test;
```
[the remark this answers is not in the document]
Comment (rusty1968) [reply test/tests/caliptra_integration_tests/main.rs]: That was unintended . Done,. I reverted all changes to tests and then added only the  "use SocManager"  to minimize changes.

File: sw-emulator/lib/types/src/mmio.rs [hunk 8]
```diff
@@ -82,29 +82,3 @@ impl<TBus: Bus> ureg::MmioMut for BusMmio<TBus> {
...
-            assert_eq!(mmio.read_volatile(4 as *const u32), 0x3abc_9321);
-            assert_eq!(mmio.read_volatile(8 as *const u16), 0x39af);
-            assert_eq!(mmio.read_volatile(10 as *const u8), 0xf3);
-        }
-        assert_eq!(
-            mmio.into_inner().data(),
-            &[0x00, 0x00, 0x00, 0x00, 0x21, 0x93, 0xbc, 0x3a, 0xaf, 0x39, 0xf3, 0x00]
-        );
-    }
-}
```
[the remark this answers is not in the document]
Comment (rusty1968) [reply sw-emulator/lib/types/src/mmio.rs]: That was unintended . Done,. I reverted all changes to tests and then added only the  "use SocManager"  to minimize changes.

File: hw-model/src/lib.rs [hunk 9]
```diff
@@ -1303,94 +1051,12 @@ mod tests {
...
-        let buf = caliptra_hw_model::mbox_read_fifo(model.soc_mbox());
-        assert_eq!(buf, &[0, 0, 0]);
-    }
-
-    #[test]
-    // Currently only possible on verilator
-    // SW emulator does not support pauser
-    // For FPGA, test case needs to be reworked to capture SIGBUS from linux environment
-    #[cfg(feature = "verilator")]
-    fn test_mbox_pauser() {
```
[the remark this answers is not in the document]
Comment (rusty1968) [reply hw-model/src/lib.rs]: That was unintended . Done,. I reverted all changes to tests and then added only the  "use SocManager"  to minimize changes.

File: hw-model/src/lib.rs [hunk 10]
```diff
@@ -1089,88 +867,58 @@ pub trait HwModel {
...
-        if response.hdr.fips_status != api::mailbox::MailboxRespHeader::FIPS_STATUS_APPROVED {
-            return Err(ModelError::UploadMeasurementResponseError);
-        }
-
-        Ok(())
+        SocManager::upload_meas(self, measurement).map_err(ModelError::from)
     }
 }
 
+pub fn wait_for_mailbox_receive<'a, 'b, TModel: HwModel + SocManager>(
```
[the remark this answers is not in the document]
Comment (rusty1968) [reply hw-model/src/lib.rs]: I removed all the mailbox apis for the time being so we can focus on the refactoring and approve this subset.

File: hw-model/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (zhalvorsen) [reply hw-model/src/lib.rs:313]: If nothing in the firmware references trying to display the error, the Rust compiler is smart enough to exclude the implementation from the embedded firmware. The  `std::Display` and `core::fmt::Display` APIs are the same so it wouldn't require any changes to the current implementation. There is also a `core` version of `Error` (`core::error::Error`) that could be brought to the `no_std` version.

[the remark this answers is not in the document]
Comment (rusty1968) [reply hw-model/src/lib.rs:313]: write! is part of std::io.

[the remark this answers is not in the document]
Comment (zhalvorsen) [reply hw-model/src/lib.rs:313]: It is also in core https://doc.rust-lang.org/core/macro.write.html

[the remark this answers is not in the document]
Comment (rusty1968) [reply hw-model/src/lib.rs:313]: > If nothing in the firmware references trying to display the error, the Rust compiler is smart enough to exclude the implementation from the embedded firmware. The `std::Display` and `core::fmt::Display` APIs are the same so it wouldn't require any changes to the current implementation. There is also a `core` version of `Error` (`core::error::Error`) that could be brought to the `no_std` version.

That's an unstable feature.
**error in core is an unstable feature. 
error[E0658]: use of unstable library feature 'error_in_core'
 --> api/src/lib.rs:3:5
  |
3 | use core::error::Error;**

[the remark this answers is not in the document]
Comment (zhalvorsen) [reply hw-model/src/lib.rs:313]: We talked about this a little offline. It looks like the feature stabilized after the version of Rust we are using in Caliptra. Lets just leave it as is and we can optimize later when upgrade the version of the compiler.

File: api/src/mailbox.rs [hunk 11]
```diff
@@ -1047,6 +1054,98 @@ pub struct AuthorizeAndStashResp {
...
+    fn deref(&self) -> &Self::Target {
+        &self.data
+    }
+}
+
+impl DerefMut for MboxBuffer {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.data
+    }
+}
```
Comment (zhalvorsen) [inline api/src/mailbox.rs:1083]: Should we remove these now that they aren't used?

File: hw-model/src/lib.rs [hunk 12]
```diff
@@ -629,7 +660,7 @@ pub trait HwModel {
                 "Fuses are already locked in place (according to cptra_fuse_wr_done)"
             );
         }
-        println!("Initializing fuses: {:#x?}", fuses);
+        //        println!("Initializing fuses: {:#x?}", fuses);
```
Comment (zhalvorsen) [inline hw-model/src/lib.rs:663]: I think we can leave this uncommented out because it is in the `std` file

**Review by zhalvorsen — DISMISSED** [review]

File: api/src/mailbox.rs (hunk 11 above)
Comment (rusty1968) [reply api/src/mailbox.rs]: Done

File: hw-model/src/lib.rs (hunk 12 above)
Comment (rusty1968) [reply hw-model/src/lib.rs]: Done

**Review by zhalvorsen — APPROVED** [review]

File: api/src/mailbox.rs [hunk 13]
```diff
@@ -1047,6 +1051,70 @@ pub struct AuthorizeAndStashResp {
...
+    }
+
+    Ok(())
+}
+
+pub fn mbox_write_fifo(
+    mbox: &mbox::RegisterBlock<impl MmioMut>,
+    buf: &[u8],
+) -> core::result::Result<(), CaliptraApiError> {
+    const MAILBOX_SIZE: u32 = 128 * 1024;
```
Comment (swenson) [inline api/src/mailbox.rs:1092]: This constant appears in about a dozen places. Can we pull it from somewhere rather than creating another constant?

File: api/src/soc_mgr.rs [hunk 14]
```diff
@@ -0,0 +1,63 @@
+// Licensed under the Apache-2.0 license
+use ureg::MmioMut;
+
```
Comment (swenson) [inline api/src/soc_mgr.rs:3]: It would be good to add doc comments for the trait and for each field and method that people need to overwrite, since this is a public API to be implemented.

File: api/types/src/lib.rs [hunk 15]
```diff
@@ -0,0 +1,208 @@
...
+            0b1100 => Ok(Self::Xc),
+            0b1101 => Ok(Self::Xd),
+            0b1110 => Ok(Self::Xe),
+            0b1111 => Ok(Self::Xf),
+            16_u32..=u32::MAX => Err(()),
+        }
+    }
+}
+
+#[derive(Clone, Copy, Debug)]
```
Comment (swenson) [inline api/types/src/lib.rs:155]: This feels like a pretty big struct to have `Copy` semantics. It feels like `Clone` would be preferred so that people don't accidentally copy it and not understand the performance penalty.

File: hw-model/src/model_verilated.rs [hunk 16]
```diff
@@ -111,10 +112,38 @@ fn ahb_txn_size(ty: AhbTxnType) -> RvSize {
...
-impl crate::HwModel for ModelVerilated {
+    fn mmio_mut(&mut self) -> Self::TMmio<'_> {
+        BusMmio::new(self.apb_bus())
+    }
+
+    fn delay(&mut self) {
+        self.step();
+    }
+
+    const SOC_IFC_ADDR: u32 = 0x3003_0000;
```
Comment (swenson) [inline hw-model/src/model_verilated.rs:126]: Might be nice to move these constants to somewhere shared just so they aren't repeated (and we are less likely to make a typo).

**Review by swenson — APPROVED** [review]
LGTM. I left some minor comments, and some (like the `Copy` in the fuses) are not necessarily in the scope for this PR since they are from code that was just moved around, and I don't know how invasive of a change it will be to fix it in this PR.

**Review by JohnTraverAmd — APPROVED** [review]

**Review by mhatrevi — APPROVED** [review]

File: api/src/mailbox.rs (hunk 13 above)
Comment (rusty1968) [reply api/src/mailbox.rs:1092]: @swenson , In my follow-up to this PR, I tried to pull in the constant from the caliptra-drivers crate. For some convoluted reason I can't pretend to understand, the hash of caliptra-rom changed.

File: api/src/soc_mgr.rs (hunk 14 above)
Comment (rusty1968) [reply api/src/soc_mgr.rs:3]: I am doing this next in part-2 of caliptra-api.

File: api/types/src/lib.rs (hunk 15 above)
Comment (rusty1968) [reply api/types/src/lib.rs:155]: Done in part-2 as well.
